### PR TITLE
Support multicast

### LINKED ISSUE #459: Enforce multicast
**Summary:** 

Enforce multicast in the Network Server.

**Why do we need this?**

To avoid downlink and make sure that MAC is not processed nor generated.

**What is already there? What do you see now?**

Through ABP sessions we implicitly support multicast.

**What is missing? What do you want to see?**

We should add enforcement of multicast rules:

1. Uplink are not allowed
2. MAC commands are not allowed
3. From 1, confirmed downlink from the application layer is not allowed

**How do you propose to implement this?**

1. Add a `multicast` field to `EndDevice`. This value cannot be changed. After identifying the device and with this flag set, the downlink should be dropped
2. We should never have a diff between desired and current parameters. We should probably skip all MAC command handling so that `FOpts` are always empty and downlink with `FPort` of 0 is never generated, but we should also enforce towards clients that desired parameters cannot be set
3. When receiving downlink traffic from the application layer for a multicast device, we should return invalid argument

**Can you do this yourself and submit a Pull Request?**

Yes, but I leave the honor to @rvolosatovs

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Closes #459
Refs #789 

#### Changes

- Add `EndDevice.SupportsUplink` (`true` by default)
- Actually update app queue if downlink is failed
- Drop join-requests from ABP devices
- Do not delete in registries if `len(paths) > 0`, delete the field values specified instead
- Only validate optional fields on create(`mac_settings`)
- Queue events as preparation for #789 

#### Notes for Reviewers

`SupportsUplink` was chosen, since that aligns best with `SupportsJoin` we have already.
We may want to add `--multicast` flag to CLI, which would set `supports_join` and `supports_uplink` to `false` (cc @htdvisser )?
Console should have `Multicast` checkbox I suppose, which would do the same. (cc @bafonins @kschiffer )

### CODE REVIEW

**Comment by rvolosatovs** [comment]
Tested on `ff1705` via `test-class-c-device ff1705 --mac-settings.status-time-periodicity 10s --mac-settings.status-count-periodicity 2`(https://github.com/rvolosatovs/lorawan-stack/blob/0d353f3a48b5fa02c9eb92e11854069bd8440eb1/shell.nix#L225-L270)

File: pkg/networkserver/grpc_asns.go [hunk 1]
```diff
@@ -182,19 +213,19 @@ func (ns *NetworkServer) DownlinkQueuePush(ctx context.Context, req *ttnpb.Downl
 
 	dev, err := ns.devices.SetByID(ctx, req.EndDeviceIdentifiers.ApplicationIdentifiers, req.EndDeviceIdentifiers.DeviceID,
 		[]string{
+			"mac_state",
+			"pending_mac_state",
+			"pending_session",
 			"queued_application_downlinks",
-			"mac_state.device_class",
```
Comment (rvolosatovs) [inline pkg/networkserver/grpc_asns.go]: If stored `mac_state` is `nil`, `dev.MACState` will be equal to `&ttnpb.MACState{}`, which causes panic in validation.

**Comment by johanstokking** [comment]
I think we should stay close to the spec and behavior and name the flag (and field) `multicast`. Multicast implies that the device doesn't support uplink, but `multicast` may mean more (like not supporting MAC commands and for the Console some multicast session management). I would not make that implicitly depend on `supports_uplink`.

File: pkg/networkserver/downlink.go [hunk 2]
```diff
@@ -96,6 +96,22 @@ type generatedDownlink struct {
 	Priority            ttnpb.TxSchedulePriority
 }
 
+type generateDownlinkState struct {
+	baseApplicationUps        []*ttnpb.ApplicationUp
+	ifScheduledApplicationUps []*ttnpb.ApplicationUp
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/downlink.go]: This is handled by https://github.com/TheThingsNetwork/lorawan-stack/pull/792/files/f409d316489307f92ee3f438974165992c070308#diff-c3a60fbddbdd11d06dbb71878862ac4bR107

`baseApplicationUps` are added to the queue if downlink is generated
`ifScheduledApplicationUps` is added to the queue if downlink is actually scheduled

Note, that `baseApplicationUps` may be discarded if Rx1 of Class B/C downlink does not apply.

**Comment by htdvisser** [comment]
@rvolosatovs 	please create an issue for CLI and Console support for enabling multicast for a device (or creating multicast devices)

**Comment by rvolosatovs** [comment]
@htdvisser I did not create an issue about CLI, since that should be just a matter of doing `--multicast=true` on device creation, which I assume should be supported already, since that's just a new field on the proto.
If anything is still missing in CLI implementation, please comment on what is it exactly that is missing.

**Comment by htdvisser** [comment]
Ok, yeah new fields should get picked up without any changes. I commented that before you edited the original PR message.

**Comment by johanstokking** [comment]
@rvolosatovs please do file an issue for updating the getting started; there’s a multicast example that needs to be updated

**Comment by rvolosatovs** [comment]
https://github.com/TheThingsNetwork/lorawan-stack/issues/850

**Pull request merged**